Staging/EnableGop/EnableGop.c

```diff
@@ -0,0 +1,231 @@
+  OcSetSystemVariable (
+    Name,
+    OPEN_CORE_NVRAM_ATTR,
+    sizeof (One),
+    &One,
+    NULL
+    );
+}
+
+STATIC
+EFI_STATUS
+LoadUefiOutputSupport (
+  VOID
+  )
+{
+  EFI_STATUS  Status;
+
+  Status = OcProvideConsoleGop (FALSE);
+  if (EFI_ERROR (Status)) {
+    LogStage ('2');
+    return Status;
+  }
+
+  LogStage ('3');
+
+  OcSetConsoleResolution (
+    0,
+    0,
+    0,
+    FALSE
+    );
+
+  if (FeaturePcdGet (PcdEnableGopDirect)) {
+    OcUseDirectGop (-1);
+  }
+
+  OcSetupConsole (
+    OcConsoleRendererBuiltinGraphics,
+    FALSE,
+    FALSE,
+    FALSE,
+    FALSE
+    );
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+ProvideGop (
+  VOID
+  )
+{
+  LogStage ('1');
+
+  OcUnlockAppleFirmwareUI ();
+
+  OcForgeUefiSupport (TRUE, TRUE);
+  OcReloadOptionRoms ();
+
+  BdsLibConnectAllDriversToAllControllers ();
+
+  return LoadUefiOutputSupport ();
+}
+
+STATIC
+EFI_STATUS
+SetDefaultVariable (
+  CHAR16    *Name,
+  EFI_GUID  *Guid,
+  UINTN     Size,
+  VOID      *Data
+  )
+{
+  EFI_STATUS  Status;
+  UINTN       DataSize;
+
+  DataSize = 0;
+  Status   = gRT->GetVariable (
+                    Name,
+                    Guid,
+                    NULL,
+                    &DataSize,
+                    NULL
+                    );
+
+  //
+  // Do not modify existing value.
+  //
+  if (Status == EFI_BUFFER_TOO_SMALL) {
+    return EFI_SUCCESS;
+  }
+
+  return gRT->SetVariable (
+                Name,
+                Guid,
+                EFI_VARIABLE_BOOTSERVICE_ACCESS | EFI_VARIABLE_RUNTIME_ACCESS | EFI_VARIABLE_NON_VOLATILE,
+                Size,
+                Data
+                );
+}
+
+//
+// If not present and large enough, Apple picker is never entered on some GPUs.
+// Gets (re-)set on first boot of macOS.
+//
+STATIC
+EFI_STATUS
+SetBluetoothDelay (
+  VOID
+  )
+{
+  UINT16  EFIBluetoothDelay;
+
+  EFIBluetoothDelay = EFI_BLUETOOTH_DELAY_DEFAULT;
+
+  return SetDefaultVariable (
+           L"EFIBluetoothDelay",
+           &gAppleBootVariableGuid,
+           sizeof (EFIBluetoothDelay),
+           &EFIBluetoothDelay
+           );
+}
+
+//
+// This memory map access happens in the equivalent of efi InitializeMemoryTest
+// at the start of PlatformBdsPolicyBehavior and just after BdsLibLoadDrivers
+// in 144.0.0.0.0 Mac Pro firmware.
+//
+EFI_STATUS
+EFIAPI
+WrappedGetMemorySpaceMap (
+  OUT UINTN                            *NumberOfDescriptors,
+  OUT EFI_GCD_MEMORY_SPACE_DESCRIPTOR  **MemorySpaceMap
+  )
+{
+  STATIC UINTN  mGetMemorySpaceMapAccessCount = 0;
+
+  mGetMemorySpaceMapAccessCount++;
+
+  if (mGetMemorySpaceMapAccessCount == 1) {
+    SetBluetoothDelay ();
+    ProvideGop ();
+  }
+
+  return mOriginalGetMemorySpaceMap (
+           NumberOfDescriptors,
+           MemorySpaceMap
+           );
+}
+
+STATIC
+EFI_STATUS
+WrapGetMemorySpaceMap (
+  VOID
+  )
+{
+  mOriginalGetMemorySpaceMap = gDS->GetMemorySpaceMap;
+  gDS->GetMemorySpaceMap     = WrappedGetMemorySpaceMap;
```

Comment:
No checksum here?

Reply:
I saw you'd added all this check-summing, but I've never seen it matter - maybe is required only on different/newer firmware? If it is required somewhere, I agree I should add it.

(I wrote tons of trapping, of gBS, gRT and this, while developing this, and never checksummed - although ... oh ... I did get one crash when trapping early ... maybe that was it!)

Reply:
This is definitely *sometimes* checked. Probably not for gDS, because this dies when exiting boot services, but definitely for gRT. This is a clear case of "just do it". :)

Reply:
I think we did check for gDS checksum in our Linux patches when we added memory protection code to the efistub. Please fix.